src/dmd/parse.d

```diff
@@ -3958,6 +3958,12 @@ final class Parser(AST) : Lexer
                     //printf("it's type[expression]\n");
                     inBrackets++;
                     AST.Expression e = parseAssignExp(); // [ expression ]
+                    if (!e)
```

Comment:
If what was between the brackets has failed to parse, `e` is going to be null causing the segfault.